### PR TITLE
[NVPTX] Fold (add (select 0, (mul a, b)), c) -> (select c, (mad a, b, c))

### PR DESCRIPTION
**Pull request by AlexMaclean**
Add folding for `(add (select cond, 0, (mul a, b)), c)` to `(select cond, c, (mad a, b, c))`. Also, refactor the DAG folding implementation to separate out the `ADD` and `FADD` folding cases.

### CODE REVIEW

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 1]
```diff
@@ -5215,103 +5215,129 @@ bool NVPTXTargetLowering::allowUnsafeFPMath(MachineFunction &MF) const {
...
-  EVT VT=N0.getValueType();
-  if (VT.isVector())
+static SDValue
+PerformADDCombineWithOperands(SDNode *N, SDValue N0, SDValue N1,
+                              TargetLowering::DAGCombinerInfo &DCI) {
+  EVT VT = N0.getValueType();
+
+  // Since integer multiply-add costs the same as integer multiply
+  // but is more costly than integer add, do the fusion only when
+  // the mul is only used in the add.
```
Comment (jlebar) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:5234]: 1. Is this statement true?  According to https://docs.nvidia.com/cuda/cuda-c-programming-guide/#arithmetic-instructions, on sm70+ integer multiply, integer add, and integer madd all have the same throughput.

3. If we really do want to add the hasOneUse() constraint for the simple madd transformation, we should definitely call it out in the commit message (or even do it as a separate commit?)

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp [hunk 2]
```diff
@@ -5215,103 +5215,129 @@ bool NVPTXTargetLowering::allowUnsafeFPMath(MachineFunction &MF) const {
...
+  if (N0.getOpcode() == ISD::SELECT) {
+    bool ZeroCond;
+    if (isConstZero(N0->getOperand(1)))
+      ZeroCond = true;
+    else if (isConstZero(N0->getOperand(2)))
+      ZeroCond = false;
+    else
+      return SDValue();
+
+    SDValue M = N0->getOperand(ZeroCond ? 2 : 1);
```
Comment (jlebar) [inline llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:5256]: I feel like you're using a boolean here to represent "operand 1 is zero or operand 2 is zero".  But maybe using an integer with values 1 or 2 would make more sense?  Like, "ZeroCond is false therefore N0->getOperand(2) is zero" doesn't make sense to me.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 1 above)
Comment (AlexMaclean) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:5234]: Both this statement and the `hasOneUse()` constraint were here before this change. I've just refactored them around a little to avoid the need for code duplication. I agree the `hasOneUse()`, may be a little too conservative, but relaxing this constrain poses some risks and I think it makes sense to maintain it as is for the purposes of this commit.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 2 above)
Comment (AlexMaclean) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp]: Sounds good, I've updated this to use an index.

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 1 above)
Comment (jlebar) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:5234]: Ah, I see it.  Sorry, I looked but didn't see the logic before.

Maybe add a TODO that this statement may not be true?

**Review by jlebar — APPROVED** [review]

File: llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp (hunk 1 above)
Comment (AlexMaclean) [reply llvm/lib/Target/NVPTX/NVPTXISelLowering.cpp:5234]: Sounds good, added a TODO
